### PR TITLE
Allow race-free implementation of the State Caching Proxy Pattern.

### LINKED ISSUE #187: Unable to correctly implement State Caching Proxy Pattern using dbusutil
Hi All,

*Background:*
I am working on an application that involves tracking the state of a DBus object. The object I am using provides a method to get its current state, as well as a signal to notify us whenever that state changes.

I would like to build a local object which tracks the state of the remote object and caches it for quick access. In the literature, I have seen this referred to as a a _State Caching Proxy pattern_. 

The guarantee I would like to achieve is that once the object has been initialised, the observable states of this local object will be a contiguous subsequence of the observable states of the remote object. That is, if the remote object transitions through states S(0), S(1) ... S(n) during the lifetime of the system, the local object will expose some contiguous subsequence of those states (for as the local object exists and is subscribed). 
Valid observation trace: S(3), S(4), S(5), S(6) ...
Invalid observation trace: S(4), S(3), S(5), S(6) ... (States received out of order.)
Invalid observation trace: S(3), S(5), S(6). (Missing state changes.)

If at the point the local caching proxy object is initialised, it exposes state S(3), I expect to be able to observe all future state changes (i.e. S(4), S(5), S(6), ..) in exactly that order without missing any changes *and* without any other state changes mixed in. 

The State Caching Proxy pattern is described in section 5.2 of Olli Salli's masters thesis:
"Building object-oriented software with the D-Bus messaging system"
https://www.utupub.fi/bitstream/handle/10024/78669/gradu2012Salli.pdf?sequence=1

As described in that thesis, the correct way of implementing this pattern is to subscribe for changes first and _then_ ask for the object's current state. Only once the current state is received, is the object considered initialised and are further state changes used.

*Problem*
Unfortunately, dbusutil presents two impediments to correctly implementing this pattern:
- Signals are not necessarily received in-order. The current implementation spins up a separate goroutine to dispatch each signal. There is no guarantee these goroutines will execute in-order (indeed, in our observation, for signals sent in close succession, the goroutines execute in arbitrary order. As such, we are receiving signals out of order.). 
- I have no way of identifying if a signal was received before or after the method reply that told us the initial state of the object. Both signals and method responses are handled in a somewhat asynchronous manner, using different channels and goroutines. Consequently, the original order of messages on the DBus is lost. This ordering is the only way in which we can ascertain if a state change signal is for an old state (prior to us querying the initial state) and should be discarded, or is for a new state.

To my knowledge, other D-Bus libraries such as GDBus, provide mechanisms to correctly implement this pattern by offering synchronous processing of D-Bus messages.

I would be open to contributing changes to fixing dbusutil to allow the above pattern, but it seems to me that it may require significant architectural changes (mostly for the second issue: exposing ordering between method responses and signals). I would therefore like to enquire what the maintainers of this package think about this issue.

Thanks in advance,
Patrick

### PR DESCRIPTION
**Pull request by patrickmeiring**
Fixes #187.

This is a prototype of my proposed solution, with tests. I am unsure about the naming of external interface members and whether additional changes should be made to the external interface (other than adding Signal.Sequence and Call.ResponseSequence).

If you are completely happy with this change, please merge, otherwise please provide feedback and I will address.

Thanks,
Patrick

### CODE REVIEW

File: conn_test.go [hunk 1]
```diff
@@ -256,6 +257,163 @@ func waitSignal(sigc <-chan *Signal, name string, timeout time.Duration) *Signal
...
+	defer conn.Close()
+
+	serviceName := srv.Names()[0]
+	messages := make(chan *Message)
+	srv.Eavesdrop(messages)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+	defer cancel()
+
+	done := make(chan bool)
```
Comment (jsouthworth) [inline conn_test.go]: You might want to consider a sync.WaitGroup here for readability.

File: conn_test.go [hunk 2]
```diff
@@ -256,6 +257,163 @@ func waitSignal(sigc <-chan *Signal, name string, timeout time.Duration) *Signal
...
+
+	conn, err := ConnectSessionBus(WithSignalHandler(NewSequentialSignalHandler()))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer conn.Close()
+
+	serviceName := srv.Names()[0]
+	messages := make(chan *Message)
+	srv.Eavesdrop(messages)
```
Comment (jsouthworth) [inline conn_test.go]: Why Eavesdrop? This isn't the normal way to process messages for servers.

Comment (patrickmeiring) [reply conn_test.go]: In this test, we have a server which exports an object with a state that clients can query over DBus and subscribe to state change notifications for. 

For this test, it is important that the server's outputs as sent over DBus (its emitted signals and method replies) provide a consistent view of the object's state. That is, each DBus message sent should reflect the state at some point in time, and that time should be monotonically increasing with each DBus message. I think this is basically Tanenbaum's "Monotonic read consistency".

If we are handling a request for the current state around the same time a few state change events are occuring, the following outputs would all be valid (S represents a state change signal, R a response to a method call querying the current state):
- S(38->39), R(39), S(39->40) 
- R(38), S(38->39), S(39->40) 
- S(38->39), S(39->40), R(40)

These, on the other hand, would not be valid:
S(38->39), S(39->40), R(39)
S(38->39), S(39->40), R(38)
S(38->39), R(40), S(39->40)
etc.

This appears to difficult to achieve if I simply export() an object to get the library handle method replies for me and use a second goroutine to emit events on a timer. 
If I use export(), each method request will end up calling a function I designate, but there is no way I can prevent the second goroutine from advancing the state and emitting another event _before_ my method response hits the dbus. Currently, the first goroutine to get to .sendMessageAndIfClosed() gets to send its message first, and this has nothing to do with which goroutine read the application-level state first.

File: conn_test.go (hunk 1 above)
Comment (patrickmeiring) [reply conn_test.go]: Thanks, good idea!

File: conn_test.go (hunk 2 above)
Comment (patrickmeiring) [reply conn_test.go]: In summary, the issue I'm working around here with evesdrop is basically the server-side counterpart to the client-side issue this patch is fixing. 

For our use case, this is not really a concern as the server is implemented in a different language and (consequently) does not use this library.

**Comment by patrickmeiring** [comment]
I've figured out why this test case is failing: it is because the messages channel I'm creating is unbuffered. Consequently, the server's inWorker() is dropping the client's request to call State() if serverProcess isn't currently waiting on <-messages (because, for example, it is busy emitting an event.)

This can be fixed by making the message channel's buffer large enough to not drop any messages. (a 1 message buffer seems sufficient in this case, assuming no messages from any other sources.)

See this snippet from conn.go:383:
```
		conn.eavesdroppedLck.Lock()
		if conn.eavesdropped != nil {
			select {
			case conn.eavesdropped <- msg:
			default:
			}
			conn.eavesdroppedLck.Unlock()
			continue
		}
		conn.eavesdroppedLck.Unlock()
``

**Comment by patrickmeiring** [comment]
Looks like there is another issue that I don't know about.

In both cases, the test is now failing at a later point:
    conn_test.go:320: Subscribed to events
    conn_test.go:396: Processing reply to .State(), returning state = 1
    conn_test.go:333: Queried current state, got 1
    conn_test.go:302: error in client process: context deadline exceeded

Whereas in the earlier case, they both failed at:
    conn_test.go:317: Subscribed to events
    conn_test.go:297: error in client process: context deadline exceeded

That is, the client thread is timing out reading signals rather than waiting for a response to its query for the initial state.

**Comment by patrickmeiring** [comment]
If you don't mind, I'm going to use this pull request as a bit of a development branch to try out a few changes and figure out what is going wrong on Power PC architectures. When I'm happy, I will squish down the final changes down to one commit and create a new pull request.

**Comment by patrickmeiring** [comment]
I've figured out what the problem is. golang's time.NewTicker seems to be broken on the ppc64le architecture, at least for anything less than second-level granularity.

On ppc64le:
```
--- PASS: TestTickerPerformance (28.00s)
    conn_test.go:544: 100µs ticker emitted 1 ticks/sec
    conn_test.go:544: 200µs ticker emitted 1 ticks/sec
    conn_test.go:544: 500µs ticker emitted 1 ticks/sec
    conn_test.go:544: 1ms ticker emitted 1 ticks/sec
    conn_test.go:544: 10ms ticker emitted 1 ticks/sec
    conn_test.go:544: 100ms ticker emitted 1 ticks/sec
    conn_test.go:544: 1s ticker emitted 0.75 ticks/sec
=== RUN   TestGetKey
```

On other architectures, e.g. 386, we get something like:
```
--- PASS: TestTickerPerformance (28.00s)
    conn_test.go:544: 100µs ticker emitted 9942.5 ticks/sec
    conn_test.go:544: 200µs ticker emitted 4998.75 ticks/sec
    conn_test.go:544: 500µs ticker emitted 2000 ticks/sec
    conn_test.go:544: 1ms ticker emitted 1000 ticks/sec
    conn_test.go:544: 10ms ticker emitted 100 ticks/sec
    conn_test.go:544: 100ms ticker emitted 10 ticks/sec
    conn_test.go:544: 1s ticker emitted 1 ticks/sec
=== RUN   TestGetKey
```

Test code:
```go
func TestTickerPerformance(t *testing.T) {
	testTicker(t, time.Microsecond*100)
	testTicker(t, time.Microsecond*200)
	testTicker(t, time.Microsecond*500)
	testTicker(t, time.Millisecond)
	testTicker(t, time.Millisecond*10)
	testTicker(t, time.Millisecond*100)
	testTicker(t, time.Millisecond*1000)
}

func testTicker(t *testing.T, d time.Duration) {
	count := 0

	ctx, cancel := context.WithTimeout(context.Background(), time.Second*4)
	defer cancel()

	ticker := time.NewTicker(d)
	defer ticker.Stop()

loop:
	for {
		select {
		case <-ticker.C:
			count++
		case <-ctx.Done():
			break loop
		}
	}
	t.Logf("%v ticker emitted %v ticks/sec", d, float64(count)/4.0)
}
```

**Comment by patrickmeiring** [comment]
Raised https://github.com/golang/go/issues/36592

**Pull request closed without merging**